modules/src/main/java/org/archive/modules/deciderules/MatchesListRegexDecideRule.java

```diff
@@ -91,7 +107,17 @@ protected boolean evaluate(CrawlURI uri) {
         boolean listLogicOR = getListLogicalOr();
 
         for (Pattern p: regexes) {
-            boolean matches = p.matcher(str).matches();
+            boolean matches = false;
+            if (getTimeoutPerRegexSeconds() <= 0) {
+                matches = p.matcher(str).matches();
+            } else {
+                CompletableFuture<Boolean> matchesFuture = CompletableFuture.supplyAsync(() -> p.matcher(str).matches());
+                try {
+                    matches = matchesFuture.get(getTimeoutPerRegexSeconds(), TimeUnit.SECONDS);
```

Comment:
My understanding is that CompletableFuture.get() with a timeout doesn't actually interrupt the long-running regex, it simply doesn't wait for it to finish before progressing. Therefore the bad regex will continue running in the background on the common ForkJoin thread pool. After the MatchRule is executed a few times all threads in the pool will be busy processing bad regexs and all newly executed regex match rules will fail with a timeout thus corrupting the result of even good regexs.

Amusingly the StackOverflow answer for the general regex-with-a-timeout problem is actually answered by @kris-sigur where he suggests Heritrix's InterruptibleCharSequence class. https://stackoverflow.com/a/910798

Unfortunately we can't simply wrap str in InterruptibleCharSequence and call `matchesFuture.cancel(true)` as CompleteFuture's cancel() is [documented](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#cancel-boolean-) as not actually interrupting:

> mayInterruptIfRunning - this value has no effect in this implementation because interrupts are not used to control processing.

Using FutureTask instead like this seems to work though:

```java
        InterruptibleCharSequence interruptible = new InterruptibleCharSequence(str);
        FutureTask<Boolean> matchesFuture = new FutureTask<>(() -> p.matcher(interruptible).matches());
        ForkJoinPool.commonPool().submit(matchesFuture);
        try {
            matchesFuture.get(getTimeoutPerRegexSeconds(), TimeUnit.SECONDS);
        } catch (TimeoutException e) {
            matchesFuture.cancel(true);
            logger.warn("Timed out after " + getTimeoutPerRegexSeconds() + " seconds waiting for '" + p + "' to match.');
        }
```

Another approach that is a little more code but may be cleaner because it doesn't use extra threads at all is also suggested in the next SO answer:

https://stackoverflow.com/a/11348374

Reply:
I prefer the FutureTask method using standard API calls rather than roll-your-own multi-threading.

Reply:
I have now updated the branch with the requested changes. - Colin